nemo-physical/src/dictionary/tuple_dv_dict.rs

```diff
@@ -0,0 +1,605 @@
+/// Type constant to signify an u64 value in a tuple
+const TUPLE_TYPE_U64: u8 = 6;
+/// Type constant to signify an f32 value in a tuple
+const TUPLE_TYPE_F32: u8 = 7;
+/// Type constant to signify an f64 value in a tuple
+const TUPLE_TYPE_F64: u8 = 8;
+/// Type constant to signify a boolean value in a tuple
+const TUPLE_TYPE_BOOL: u8 = 9;
+/// Type constant to signify a 32bit dictionary id of a label (at the start of tuples)
+const TUPLE_TYPE_LABELID32: u8 = 10;
+/// Type constant to signify a 64bit dictionary id of a label (at the start of tuples)
+const TUPLE_TYPE_LABELID64: u8 = 11;
+
+crate::dictionary::bytes_buffer::declare_bytes_buffer!(
+    TupleBytesPairDictBytesBuffer,
+    TUPLE_BYTES_PAIR_DICT_BYTES_BUFFER
+);
+/// Type for byte pairs dictionary used for encoding tuples
+pub(crate) type TupleBytesPairDictionary =
+    GenericRankedPairDictionary<TupleBytesPairDictBytesBuffer>;
+
+/// Dictionary for storing IDs of tuples. This is not a subdictionary like the others, since
+/// we need mut access to the parent [MetaDvDictionary] to recursively create IDs for subterms
+/// of a tuple.
+#[derive(Debug, Default)]
+pub(crate) struct TupleDvDict {
+    /// Internal dictionary to store tuple term types and parameter values.
+    ///
+    /// Tuple values are encoded in two byte sequences, *type* and *values*, as follows.
+    /// The *values* are simply a sequence of binary encodings of all parameter values,
+    /// each in some format. Values of some basic types, such as numbers, are encoded directly,
+    /// whereas values of other types are represented by dictionary ids. To decode this
+    /// blob of bytes (with non-uniform boundaries between the bytes of values), the *type*
+    /// information is needed.
+    ///
+    /// The *type* encodes the encoding types of each value, and the label of the tuple (if any).
+    /// If a label is present, then the first byte in *types* is [TUPLE_TYPE_LABELID32] or
+    /// [TUPLE_TYPE_LABELID64], depending on how large the label's dictionary ID is. This is followed
+    /// by the (4 or 8) bytes of the label. Thereafter is a list of bytes that encode the types of
+    /// the values at the corresponding position. In many cases, there are as many type bytes as
+    /// values in the *values* sequence. However, repeated occurrences of the same type at the very
+    /// end of the sequence are omitted. This allows in particular to encode uniformly typed lists of
+    /// any length to share the same *type*.
+    dict: TupleBytesPairDictionary,
+}
+
+impl TupleDvDict {
+    /// Construct a new and empty dictionary.
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl DvDict for TupleDvDict {
+    /// This method always rejects.
+    fn add_datavalue(&mut self, _dv: AnyDataValue) -> AddResult {
+        AddResult::Rejected
+    }
+
+    fn add_datavalue_with_parent_fn(
+        &self,
+    ) -> fn(&mut MetaDvDictionary, dict_id: usize, dv: AnyDataValue) -> AddResult {
```

Comment:
Same closure idea also applies here

Reply:
I did this now for the short default implementations in datavalue_dictionary.rs. I am not convinced that the same approach would make the longer code in tuple_dv_dict.rs more readable though. At least one of the callbacks there also has own documentation that describes when it will panic; such docs would probably not fit in front of a closure, and having this all indented by several more levels seems not so readable to me.